IRGen: make Windows use resilient conformances

compnerd opened the pull request:
Use the resilient conformance code path for PE/COFF as that forces the witnesses
to be generated at runtime avoiding the cross-module initialised static data.

<!-- What's in this pull request? -->
Replace this paragraph with a description of your changes and rationale. Provide links to external references/discussions if appropriate.

<!-- If this pull request resolves any bugs in the Swift bug tracker, provide a link: -->
Resolves [SR-NNNN](https://bugs.swift.org/browse/SR-NNNN).

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

slavapestov commented on lib/IRGen/GenProto.cpp:
```diff
@@ -1276,21 +1279,22 @@ class AccessorConformanceInfo : public ConformanceInfo {
-                                  : ProtocolInfoKind::Full))) {
+                                      ? ProtocolInfoKind::RequirementSignature
+                                      : ProtocolInfoKind::Full))) {
       // If the conformance is resilient, we require runtime instantiation.
```
I don't think you need any of the other changes in this PR. Instead, just add a check for the triple here and set ResilientConformance to true.

Also just as with class instantiation I would prefer it if you added a new flag to IRGenOptions instead of checking the target here directly.

compnerd replied:
Hmm, okay, that does simplify things.  Unfortunately, when building the standard library itself is where things get interesting, and what I'm trying to figure out still.  It seems that `getFunctionIndex` is asserting at `getKind() >= ProtocolInfoKind::Full` since we are now just doing the `RequirementSignature`

slavapestov replied:
Sorry, I meant ‘or’ the additional condition when you set the ResilientConformance flag and not when you use it.

compnerd commented:
@swift-ci please test

compnerd committed:
IRGen: avoid constant conformances on Windows

Windows does not permit cross-module data accesses to be direct.  This
is a problem for public protocols with root conformances which are
external.  Use a runtime initialiser for the root protocol conformance
chaining to alleviate this issue.  This shows up in the Foundation
build.

compnerd commented:
@swift-ci please test

compnerd commented:
@swift-ci please test

compnerd commented:
@swift-ci please test

compnerd commented:
ping @slavapestov @rjmccall

rjmccall commented on lib/Frontend/CompilerInvocation.cpp:
```diff
@@ -1075,6 +1075,7 @@ static bool ParseIRGenArgs(IRGenOptions &Opts, ArgList &Args,
   // PE/COFF cannot deal with the cross-module reference to the metadata parent
   // (e.g. NativeObject).  Force the lazy initialization of the VWT always.
   Opts.LazyInitializeClassMetadata = Triple.isOSBinFormatCOFF();
+  Opts.LazyInitializeProtocolConformances = Triple.isOSBinFormatCOFF();
```
Please add a comment explaining why this is necessary.

compnerd replied:
Sure, done.

compnerd commented:
@swift-ci please test

rjmccall commented:
Sorry, I should've left this comment before, but: test case?

compnerd commented:
Haven't been able to reduce one, I've been testing by building Foundation.

compnerd commented:
@swift-ci please test

compnerd commented:
@swift-ci please test

compnerd commented:
@swift-ci please test

compnerd commented:
@swift-ci please test

compnerd commented:
@swift-ci please test

compnerd commented:
@swift-ci please test macOS platform

compnerd commented:
@swift-ci please test

compnerd commented:
@swift-ci please test Linux platform

compnerd commented:
@swift-ci please test

compnerd commented:
@swift-ci please test

compnerd commented:
@swift-ci please test

compnerd commented:
@swift-ci please test

compnerd commented:
@swift-ci please test macOS platform

compnerd commented:
@swift-ci please test

compnerd commented:
@swift-ci please test

compnerd commented:
@rjmccall - okay, this also now has a test.  Anything else, or is this okay to merge now?
